pgtransaction/atomic.py

```diff
@@ -58,38 +59,40 @@ def __call__(self, func):
 
         @wraps(func)
         def inner(*args, **kwds):
-            inst = self._recreate_cm()
-            with inst:
+            num_retries = 0
+
+            while True:  # pragma: no branch
                 try:
-                    return func(*args, **kwds)
-                except (
-                    psycopg2.errors.SerializationFailure,
-                    psycopg2.errors.DeadlockDetected,
-                ) as error:
-                    if self.retry > 0:
-                        self.retry -= 1
-                        # Apply __exit__ to rollback and clean up the
-                        # failed transaction correctly
-                        self.__exit__(
+                    with self._recreate_cm():
+                        return func(*args, **kwds)
+                except Error as error:
+                    if (
+                        error.__cause__.__class__
+                        not in (
                             psycopg2.errors.SerializationFailure,
-                            error,
-                            "",
+                            psycopg2.errors.DeadlockDetected,
                         )
-                    else:
+                        or num_retries >= self.retry
+                    ):
                         raise
```

Comment:
If we raise a SerializationFailure or DeadlockDetected error here, aren't we leaving the data from the transaction where those errors were invoked in an unfinished state? What's rolling back the data correctly if we don't manually invoke __exit__ as was done previously?

Reply:
`with self._recreate_cm()` is inside of the try/except, so the transaction is rolled back before we continue the retry loop

Reply:
Ahh yes of course. This way makes much more sense than the original.